File: app/src/main/java/com/luca020400/amt/Parser.java
package com.luca020400.amt;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.util.LinkedList;
import java.util.List;

class Parser {
    private final String url;
    private final String code;

    Parser(String url, String code) {
        this.url = url;
        this.code = code;
    }

    List<Stop> parse() throws IOException {
        List<Stop> stops = new LinkedList<>();

        Document document = Jsoup.connect(url).data("CodiceFermata", code).get();
        Elements trs = document.select("tr");
        trs.remove(0);

        for (Element tr : trs) {
            Elements tds = tr.select("td");
            if (tds.size() == 4) {
                stops.add(new Stop(tds.get(0).text(), tds.get(1).text(), tds.get(2).text(), tds.get(3).text()));
            }
        }
        return stops;
    }
}

File: app/src/main/kotlin/com/luca020400/amt/Parser.kt
package com.luca020400.amt

import android.util.Log
import org.jsoup.Jsoup
import org.jsoup.nodes.Document

internal class Parser(private val url: String, private val code: String) {
    private val sTAG = "Parser"

    fun parse(): Stop {
        val document: Document

        try {
            document = Jsoup.connect(url).data(Constants.query, code).get()
            if (!document.hasText()) {
                throw Throwable("Document is empty")
            }
        } catch (e: Throwable) {
            Log.e(sTAG, e.message, e)
            return Stop(code, null, listOf())
        }

        val name = document.select("font")[1].text()

        return Stop(code, name, document.select("tr")
                .map { it.select("td") }
                .filter { it.size == 4 }
                .map { StopData(it[0].text(), it[1].text(), it[2].text(), it[3].text()) })
    }
}
